src/Components/WebView/WebView/src/WebViewManager.cs

```diff
@@ -0,0 +1,184 @@
+            }
+
+            // If the page is already attached, remove the root component from it now. Otherwise it's
+            // enough to have updated the dictionary.
+            return _currentPageContext != null
+                ? _currentPageContext.Renderer.RemoveRootComponentAsync(selector)
+                : Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Notifies the <see cref="WebViewManager"/> about a message from JavaScript running within the web view.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        protected void MessageReceived(Uri sourceUri, string message)
+        {
+            if (!_appBaseUri.IsBaseOf(sourceUri))
+            {
+                // It's important that we ignore messages from other origins, otherwise if the webview
+                // navigates to a remote location, it could send commands that execute locally
+                return;
+            }
+
+            _ = _dispatcher.InvokeAsync(async () =>
+            {
+                // TODO: Verify this produces the correct exception-surfacing behaviors.
+                // For example, JS interop exceptions should flow back into JS, whereas
+                // renderer exceptions should be fatal.
+                try
+                {
+                    await _ipcReceiver.OnMessageReceivedAsync(_currentPageContext, message);
+                }
+                catch (Exception ex)
+                {
+                    _ipcSender.NotifyUnhandledException(ex);
+                    throw;
+                }
+            });
+        }
+
+        protected bool TryGetResponseContent(string uri, bool isPageLoad, out int statusCode, out string statusMessage, out Stream content, out string headers)
+            => _staticContentProvider.TryGetResponseContent(uri, isPageLoad, out statusCode, out statusMessage, out content, out headers);
+
+        internal async Task AttachToPageAsync(string baseUrl, string startUrl)
+        {
+            // If there was some previous attached page, dispose all its resources. TODO: Are we happy
+            // with this pattern? The alternative would be requiring the platform author to notify us
+            // when the webview is navigating away so we could dispose more eagerly then.
+            _currentPageContext?.Dispose();
+
+            var serviceScope = _provider.CreateScope();
+            _currentPageContext = new PageContext(_dispatcher, serviceScope, _ipcSender, baseUrl, startUrl);
+
+            // Add any root components that were registered before the page attached
+            foreach (var (selector, rootComponent) in _rootComponentsBySelector)
+            {
+                await _currentPageContext.Renderer.AddRootComponentAsync(
+                    rootComponent.ComponentType,
+                    selector,
+                    rootComponent.Parameters);
```

Comment:
I'm not sure about automatically re-rendering the components in this situation.

There's a chance that rootComponent.Parameters contains objects that have been mutated and given that we might go to a different host page, there is a chance that the same root components might not apply to the new page?

Reply:
Do you have more of an end-to-end scenario in mind that would be a good example of this?

I would expect in most cases people won't use the root-component parameters at all, and in the cases where they do, it's so that different host pages can have different global settings, or so that you could have a single root component shared across Blazor WebAssembly/Server/WebView and pass them some info or flags that differ based on the hosting environment. In these scenarios it wouldn't be something you'd try to mutate, and if you did try and nothing happened, that seems like it would be fine.

> given that we might go to a different host page

As it stands, the `BlazorWebView` controls only let you configure a single host page per control, so the scenario doesn't apply.